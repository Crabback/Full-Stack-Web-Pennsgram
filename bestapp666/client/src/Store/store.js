import { configureStore } from '@reduxjs/toolkit'
// import usersReducer from '../features/users/usersSlice'
// import postsReducer from '../features/posts/postsSlice'
// import commentsReducer from '../features/comments/commentsSlice'
import currentUserReducer from '../pages/LoginPage/currentUserSlice'

export default configureStore({
  reducer: {
    currentUser: currentUserReducer,
    
  }
})