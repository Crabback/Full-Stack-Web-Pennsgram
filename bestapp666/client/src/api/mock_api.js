/* istanbul ignore file */

import axios from 'axios';

//mockAPI URL 
// const rootURL = "";
//JSON-server URL
const rootURL ='http://localhost:8080/Users';
// Sends a Get request to the /Users endpoint
// returns all the users in the DB
export const getUsers = async () =>{
    try{
        const response = await axios.get(`${rootURL}`);
        return response.data;
        // the data is stored in the mockData
        // field of the response
    }
    catch(err){
        console.error(err);

    }
}

// Takes the username of a user as input
// and sends a Get request to the /user:id endpoint
// returns the attributes of the user
export const getUser = async (username) =>{
    try{    
        const response = await axios.get(`${rootURL}?username=${username}`);
        //get the fetched data's username
        let fetchedUsername;
        response.data.forEach(element => {
            fetchedUsername = element.username;
        });
        console.log(`successfully getUser by username: object with username: ${fetchedUsername}`);
        return response.data;
    }
    catch(err){
        console.error(err);
    }
}

// Takes the list of usernames as input
// and sends a Get request to the /user:id endpoint
// returns a list of user objects
export const getUsersAsList = async (usernames) =>{
    try{    
        const userObjects = [];
        const roster = await getUsers(); 
        roster.forEach((element) => {
            if(usernames.includes(element.username)){
                userObjects.push(element);
            }
        })
        userObjects.forEach((o)=>{console.log(o.username);})

        return userObjects;
    }
    catch(err){
        console.error(err);
    }
}


// Takes a user (without the id) as input
// and sends a POST request to the /user endpoint
// returns the attributes of the user with the id
export const createNewUser = async (userObject) =>{
    try{    
        const response = await axios.post(`${rootURL}`, userObject);
        console.log(`username=${userObject.username}&password=${userObject.password}&followings=${userObject.followings}&followers=${userObject.followers}&avatar=${userObject.avatar}&posts=${userObject.posts}`);
        return response.data; 
        // return the data with the id of the user
    }
    catch(err){
        console.error(err);
    }
}

export const createNewPost = async (username, postObject) =>{
    try{    
        const response = await axios.get(`${rootURL}?username=${username}`);
        let user = response.data[0];
        user.posts.push(postObject);
        const responsePut = await axios.put(`${rootURL}/${user.id}`, user);
        console.log(`successfully ${username} creates a new post`);
        return responsePut.data[0]; 
        // return the data with the id of the user
    }
    catch(err){
        console.error(err);
    }
}

export const followUser = async (username1, username2) =>{
    try{    
        const response1 = await axios.get(`${rootURL}?username=${username1}`);
        const response2 = await axios.get(`${rootURL}?username=${username2}`);
        let user1 = response1.data[0];
        let user2 = response2.data[0];
        user1.followings.push(username2);
        user2.followers.push(username1);
        const responseReturn = await axios.put(`${rootURL}/${user1.id}`, user1);
        await axios.put(`${rootURL}/${user2.id}`, user2);
        //get the fetched data's username
        console.log(`successfully ${username1} follows ${username2}`);
        return responseReturn.data;
    }
    catch(err){
        console.error(err);
    }
}

export const unfollowUser = async (username1, username2) =>{
    try{    
        const response1 = await axios.get(`${rootURL}?username=${username1}`);
        const response2 = await axios.get(`${rootURL}?username=${username2}`);
        let user1 = response1.data[0]
        let user2 = response2.data[0]
        user1.followings = user1.followings.filter(n => n !== username2);
        user2.followers = user2.followers.filter(n => n !== username1);
        const responseReturn = await axios.put(`${rootURL}/${user1.id}`, user1);
        await axios.put(`${rootURL}/${user2.id}`, user2);
        //get the fetched data's username
        console.log(`successfully ${username1} unfollows ${username2}`);
        return responseReturn.data;
    }
    catch(err){
        console.error(err);
    }
}