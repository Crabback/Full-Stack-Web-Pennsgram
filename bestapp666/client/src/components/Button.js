import "../Styles.css";
import React from 'react';

export function Button(props) {
    return (
    <button > {props.text} </button>);
  }
