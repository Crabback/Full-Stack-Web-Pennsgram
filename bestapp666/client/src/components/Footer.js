import "../Styles.css";
import React from 'react';

export function Footer() {
    return (
    <div className="footer" >
        <p className="footer_elements">Terms And Policies</p>
        <p className="footer_elements">All Copy Right Reserved</p>
    </div>);
  }
