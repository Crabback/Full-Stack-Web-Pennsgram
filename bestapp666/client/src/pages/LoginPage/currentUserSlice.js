import { createSlice } from '@reduxjs/toolkit'

export const currentUserSlice = createSlice({
  name: 'currentUser',
  initialState: {
    id: -1,
    username: '',
    password: '',
    followings: [],
    posts: []
  },
  reducers: {
    addLoginUser: (state,action) => {
      // Redux Toolkit allows us to write "mutating" logic in reducers. It
      // doesn't actually mutate the state because it uses the immer library,
      // which detects changes to a "draft state" and produces a brand new
      // immutable state based off those changes
      state =  action.user
    },
    logoutAction: state => {
      state = {
        id: -1,
        username: '',
        password: '',
        followings: [],
        posts: []
      }
    },
  }
})

export const { addLoginUser, logoutAction } = currentUserSlice.actions

export default currentUserSlice.reducer