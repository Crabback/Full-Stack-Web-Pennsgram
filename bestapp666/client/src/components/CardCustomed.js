import "../Styles.css";
import React, { useState } from 'react';
import 'bootstrap/dist/css/bootstrap.min.css';
import Form from 'react-bootstrap/Form';
import Button from 'react-bootstrap/Button';
import {LikeButton} from './LikeButton';
import ButtonGroup from 'react-bootstrap/ButtonGroup';
import Card from 'react-bootstrap/Card';
import Container from 'react-bootstrap/Container';
import Navbar from 'react-bootstrap/Navbar';
import ReactRoundedImage from "react-rounded-image";
import { NavLink } from "react-router-dom";
import { useNavigate } from "react-router-dom";
import { MentionsInput, Mention } from 'react-mentions'
import { addComment} from "../api/mock_api";
import { Nav } from 'react-bootstrap';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Modal from 'react-bootstrap/Modal';
import Stack from 'react-bootstrap/Stack';


function MyCommentsModal(props) {
  const comments = props.post.comments;
  const commentsList = comments.length === 0 ? []: comments.map((c)=>{
    return (
      <>
        <div className="bg-light border">
          <Row>
            <Col><h4>{c.author}</h4></Col>
            <Col ><p>{c.comment + " "} 
            <NavLink to={"/user/"+c.mention.replace(')', '').split("(")[1]} className="button_text">
                    {c.mention.split("(")[0]}
                </NavLink>
            </p></Col>
          </Row>
            </div>
      </>
    )
  }) ;

  return (
    <Modal
      {...props}
      size="xl"
      aria-labelledby="contained-modal-title-vcenter"
      centered
    >
      <Modal.Header closeButton>
        <Modal.Title id="contained-modal-title-vcenter">
          Comments
        </Modal.Title>
      </Modal.Header>
      <Modal.Body>
          <Stack gap={3}>
              {commentsList}
          </Stack>
        
      </Modal.Body>
      <Modal.Footer>
        <Button onClick={props.onHide}>Close</Button>
      </Modal.Footer>
    </Modal>
  );
}

function MyLikesModal(props) {
  const likes = props.post.likes;
  const likesList = likes.length === 0 ? []: likes.map((l)=>{
    return (
      <>
        <div className="bg-light border">
          <Row>
            <Col><h4>{l}</h4></Col>
          </Row> 
            </div>
      </>
    )
  }) ;

  return (
    <Modal
      {...props}
      size="xl"
      aria-labelledby="contained-modal-title-vcenter"
      centered
    >
      <Modal.Header closeButton>
        <Modal.Title id="contained-modal-title-vcenter">
          Like
        </Modal.Title>
      </Modal.Header>
      <Modal.Body>
          <Stack gap={3}>
              {likesList}
          </Stack>
        
      </Modal.Body>
      <Modal.Footer>
        <Button onClick={props.onHide}>Close</Button>
      </Modal.Footer>
    </Modal>
  );
}


export function CardCustomed(props) {
    const navigate = useNavigate();
    const [modalShow, setModalShow] = useState(false);
    const [likeModalShow, setlikeModalShow] = useState(false);

    const [descInput, setDescInput] = useState("");
    const [mentionInput, setMentionInput] = useState("");
    const [offset, setOffset] = useState(0);
    const [offsetComment, setOffsetComment] = useState(0);

    const handleLeaveComment= async (e) => {
      e.preventDefault();
      if(props.username === "NOT_A_USER"){
        alert("Please sign up/login to your account in order to make a post.");
        navigate("/");
      }else{
        const newComment = {
          "author": props.username,
          "comment": descInput,
          "mention": mentionInput
        }
        await addComment(props.post.id, newComment);
        setOffsetComment(offsetComment+1);
        alert("commented successful! re-enter this page to see the change.");
      }
      setDescInput('');
      setMentionInput('');
    }

      if (props.post) {
          return (
            <>
            <Card bg = "light" style={{ width: '28rem'}}>
              <Navbar bg="dark" variant="dark">
                <Container>
                  <Nav bg="dark" className="justify-content-end" activeKey="/">
                    <Nav.Item>
                      <NavLink to={'/user/'+props.post.author}>
                        <ReactRoundedImage image={props.post.avatar} roundedColor="#ffffff" imageWidth="35" imageHeight="35" roundedSize="2" borderRadius="20" />
                      </NavLink>
                    </Nav.Item>
                    <Nav.Item> <NavLink to={'/user/'+props.post.author} className='headers'><strong>{props.post.author}</strong></NavLink> </Nav.Item>
                  </Nav>
                </Container>
              </Navbar>
  
              {(props.post.image.split(".").slice(-1) == 'mp4') &&
            <video width='446' controls > <source src={props.post.image} type="video/mp4"/> </video>  
            }

            {(props.post.image.split(".").slice(-1) != 'mp4') &&
             <>
                <Card.Img variant="bottom" rounded="true" src={props.post.image} />
                <Card.Subtitle className="text-muted" style={{paddingTop: "1rem", paddingLeft:"1rem"}}> {props.post.date} </Card.Subtitle>
             </>
            }

              <Card.Body>
                <Card.Text>{props.post.description} </Card.Text>
                
                <Row>
                  <Col>
                  <Button variant="light" onClick={() => setlikeModalShow(true) }><Card.Text> {props.post.likes.length + offset} likes </Card.Text></Button>
                  </Col>
                  <Col>
                  <Button variant="light" onClick={() => setModalShow(true) }> <Card.Text> {props.post.comments.length + offsetComment} comments </Card.Text> </Button>
                  </Col>
                </Row>
  
                <ButtonGroup aria-label="like,comment,message">
                    <LikeButton post = {props.post} username={props.username} setOffset={setOffset}/>
                </ButtonGroup>
  
                <Form onSubmit={handleLeaveComment}>
                  <Row>
                    <Form.Group className="mb-3" controlId="formBasicDescription">
                      <Form.Control as="textarea" onChange={e => setDescInput(e.target.value)}
                      placeholder="Leave a comment." 
                      rows={2}
                      value={descInput}/>
                    </Form.Group>
                  </Row>
                  <Row style={{paddingBottom: '1rem', paddingLeft: '1rem', paddingRight: '1rem'}}>
                    <MentionsInput value={mentionInput} onChange={e => setMentionInput(e.target.value)} placeholder={"Mention people using '@'"}>
                        <Mention
                          trigger="@"
                          data={(search) => [{ id: search, display: search }]}
                        />
                      </MentionsInput>
                  </Row>
                  <Row>
                    <Button variant="primary" type="submit">
                      Comment
                    </Button>
                  </Row>
                </Form>
              </Card.Body>
            </Card>

        <MyCommentsModal
        show={modalShow}
        onHide={() => setModalShow(false)}
        post={props.post} />

        <MyLikesModal
        show={likeModalShow}
        onHide={() => setlikeModalShow(false)}
        post={props.post}
        />

        </>
            )
        
      }
  
  }
