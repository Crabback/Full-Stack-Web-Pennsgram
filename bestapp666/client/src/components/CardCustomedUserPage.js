import "../Styles.css";
import React, { useState } from 'react';
import 'bootstrap/dist/css/bootstrap.min.css';
import Form from 'react-bootstrap/Form';
import Button from 'react-bootstrap/Button';
import {LikeButton} from './LikeButton';
import ButtonGroup from 'react-bootstrap/ButtonGroup';
import Card from 'react-bootstrap/Card';
import Container from 'react-bootstrap/Container';
import Navbar from 'react-bootstrap/Navbar';
import ReactRoundedImage from "react-rounded-image";
import { NavLink } from "react-router-dom";
import { useNavigate } from "react-router-dom";
import { MentionsInput, Mention } from 'react-mentions'
import { addComment} from "../api/mock_api";
import { Nav } from 'react-bootstrap';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import OverlayTrigger from 'react-bootstrap/OverlayTrigger';
import Popover from 'react-bootstrap/Popover';
import Modal from 'react-bootstrap/Modal';
import Stack from 'react-bootstrap/Stack';
import { getUserAvatar } from "../api/mock_api";

const popover = (
  <Popover id="popover-basic">
    <Popover.Header as="h3">Popover right</Popover.Header>
    <Popover.Body>
      And here's some <strong>amazing</strong> content. It's very engaging.
      right?
    </Popover.Body>
  </Popover>
);

function MyCommentsModal(props) {
  console.log("props: shi : ", props.post);
  const comments = props.post.comments;
  const commentsList = comments.length === 0 ? []: comments.map((c)=>{
    return (
      <>
        <div className="bg-light border">
          <Row>
            <Col><h4>{c.author}</h4></Col>
            <Col ><p>{c.comment}</p></Col>
          </Row>
                
              
            </div>
      </>
    )
  }) ;

  return (
    <Modal
      {...props}
      size="xl"
      aria-labelledby="contained-modal-title-vcenter"
      centered
    >
      <Modal.Header closeButton>
        <Modal.Title id="contained-modal-title-vcenter">
          Comments
        </Modal.Title>
      </Modal.Header>
      <Modal.Body>
          <Stack gap={3}>
              {commentsList}
          </Stack>
        
      </Modal.Body>
      <Modal.Footer>
        <Button onClick={props.onHide}>Close</Button>
      </Modal.Footer>
    </Modal>
  );
}

export function CardCustomedUserPage(props) {
    const [modalShow, setModalShow] = useState(false);
    const navigate = useNavigate();
    const [descInput, setDescInput] = useState("");
    const [mentionInput, setMentionInput] = useState("");
  
    const handleLeaveComment= async (e) => {
      e.preventDefault();
      if(props.username === "NOT_A_USER"){
        alert("Please sign up/login to your account in order to make a post.");
        navigate("/");
      }else{
        const newComment = {
          "author": props.username,
          "comment": descInput,
          "mention": mentionInput
        }
        await addComment(props.post.id, newComment);
      }
      setDescInput('');
      setMentionInput('');
    }
  
      if (props.post) {
        if(props.post.image.split(".").slice(-1) == 'mp4'){
          return (
            <>
            <Card bg = "light" style={{ width: '24rem'}}>

              <video width='446' controls > <source src={props.post.image} type="video/mp4"/> </video>  
             
              <Card.Subtitle className="text-muted" style={{paddingTop: "1rem", paddingLeft:"1rem"}}> 
                {props.post.date}
              </Card.Subtitle>
    
              <Card.Body>
                <Card.Text>{props.post.description} </Card.Text>
                
                <Row style={{paddingBottom: "1rem"}}>
                  <Col> 
                        <OverlayTrigger trigger="click" placement="right" overlay={popover}>
                          <Button variant="light"><Card.Text > {props.post.likes.length} likes </Card.Text> </Button>
                        </OverlayTrigger>
                    
                  </Col>
                  <Col> <Button variant="light" onClick={() => setModalShow(true) }> <Card.Text> {props.post.comments.length} comments </Card.Text></Button> </Col>
                </Row>
  
                <ButtonGroup aria-label="like,comment,message">
                    <LikeButton post = {props.post} username={props.username}/>
                    <Button className="mb-2" variant="outline-primary">Comment</Button>
                    <Button className="mb-2" onClick={() => {
                      props.setVisibility(true);
                      props.setPostBeingEdited(props.post);
                      }}>{"edit"}</Button>
                </ButtonGroup>
  
                <Form onSubmit={handleLeaveComment}>
                  <Row>
                    <Form.Group className="mb-3" controlId="formBasicDescription">
                      <Form.Control as="textarea" onChange={e => setDescInput(e.target.value)}
                      placeholder="Leave a comment." 
                      rows={2}
                      value={descInput}/>
                    </Form.Group>
                  </Row>
                  <Row style={{paddingBottom: '1rem', paddingLeft: '1rem', paddingRight: '1rem'}}>
                    <MentionsInput value={mentionInput} onChange={e => setMentionInput(e.target.value)} placeholder={"Mention people using '@'"}>
                        <Mention
                          trigger="@"
                          data={(search) => [{ id: search, display: search }]}
                        />
                      </MentionsInput>
                  </Row>
                  <Row>
                    <Button variant="primary" type="submit">
                      Comment
                    </Button>
                  </Row>
                </Form>
  
              </Card.Body>
    
            </Card>
             {/* //pop up modal */}

             <MyCommentsModal
             show={modalShow}
             onHide={() => setModalShow(false)}
             post={props.post}
             />
         </>
            )
        }
        else{
          return (
          <>
            <Card bg = "light" style={{ width: '24rem'}}>
    
              <Card.Img variant="bottom" rounded="true" src={props.post.image} />
              <Card.Subtitle className="text-muted" style={{paddingTop: "1rem", paddingLeft:"1rem"}}> {props.post.date} </Card.Subtitle>
    
              <Card.Body>
                <Card.Text>{props.post.description} </Card.Text>
                
                <Row style={{paddingBottom: "1rem"}}>
                  <Col> 
                        <OverlayTrigger trigger="click" placement="right" overlay={popover}>
                          <Button variant="light"><Card.Text > {props.post.likes.length} likes </Card.Text> </Button>
                        </OverlayTrigger>
                    
                  </Col>
                  <Col> <Button variant="light" onClick={() => setModalShow(true) }> <Card.Text> {props.post.comments.length} comments </Card.Text></Button> </Col>
                </Row>
  
                <ButtonGroup aria-label="like,comment,message">
                    <LikeButton post = {props.post} username={props.username}/>
                    <Button className="mb-2" variant="outline-primary" onClick={() => {
                      props.setPostBeingEdited(props.post);
                      props.setVisibility(true);
                      props.setEditPanel(false);
                      }}>Comment</Button>
                    <Button className="mb-2" onClick={() => {
                      props.setPostBeingEdited(props.post);
                      props.setVisibility(true);
                      props.setEditPanel(true);
                      }}>{"edit"}</Button>
                </ButtonGroup>
  
                <Form onSubmit={handleLeaveComment}>
                  <Row>
                    <Form.Group className="mb-3" controlId="formBasicDescription">
                      <Form.Control as="textarea" onChange={e => setDescInput(e.target.value)}
                      placeholder="Leave a comment." 
                      rows={2}
                      value={descInput}/>
                    </Form.Group>
                  </Row>
                  <Row style={{paddingBottom: '1rem', paddingLeft: '1rem', paddingRight: '1rem'}}>
                    <MentionsInput value={mentionInput} onChange={e => setMentionInput(e.target.value)} placeholder={"Mention people using '@'"}>
                        <Mention
                          trigger="@"
                          data={(search) => [{ id: search, display: search }]}
                        />
                      </MentionsInput>
                  </Row>
                  <Row>
                    <Button variant="primary" type="submit">
                      Comment
                    </Button>
                  </Row>
                </Form>
              </Card.Body>
            </Card>

            {/* //pop up modal */}

            <MyCommentsModal
            show={modalShow}
            onHide={() => setModalShow(false)}
            post={props.post}
            />
        </>
            )
        }
      }
  
  }
