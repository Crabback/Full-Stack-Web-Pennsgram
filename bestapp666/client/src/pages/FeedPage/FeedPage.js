import React, { useState, useEffect } from 'react';
import 'bootstrap/dist/css/bootstrap.min.css';
import Button from 'react-bootstrap/Button';
import ButtonGroup from 'react-bootstrap/ButtonGroup';
import Card from 'react-bootstrap/Card';
import Stack from 'react-bootstrap/Stack';
import Container from 'react-bootstrap/Container';
import Navbar from 'react-bootstrap/Navbar';
import ReactRoundedImage from "react-rounded-image";
import { NavLink } from "react-router-dom";
import { useSelector} from 'react-redux'
import {selectCurrentUser} from '../UserPage/currentUserSlice'
import { getUsersAsList, getLastestPostOfAUser, getPost } from "../../api/mock_api";
import { Nav } from 'react-bootstrap';

//define a inside components
function CardCustomed(props){
    //props is a post fed from a following user
    if (props.post) {
      if(props.post.image.split(".").slice(-1) == 'mp4'){
        return (
          <Card bg = "light" style={{ width: '28rem'}}>
            <Navbar bg="dark" variant="dark">
              <Container>
                <Nav bg="dark" className="justify-content-end" activeKey="/">
                  <Nav.Item> <NavLink to={'/user/'+props.post.author}>
                      <ReactRoundedImage
                        image={props.post.avatar}
                        roundedColor="#ffffff" imageWidth="35" imageHeight="35" roundedSize="2" borderRadius="20"/>
                    </NavLink> </Nav.Item>
                  <Nav.Item> <NavLink to={'/user/'+props.post.author} className='headers'><strong>{props.post.author}</strong></NavLink> </Nav.Item>
                </Nav>
              </Container>
            </Navbar>
            <video width='446' controls autoPlay={true}> <source src={props.post.image} type="video/mp4"/> </video>  
            <Card.Subtitle className="text-muted" style={{paddingTop: "1rem"}}> 
              {props.post.date}
            </Card.Subtitle>
  
            <Card.Body>
              <Card.Text> {props.post.description} </Card.Text>
              <ButtonGroup aria-label="like,comment,message">
                  <Button variant="light">Like</Button>
                  <Button variant="light">Comment</Button>
              </ButtonGroup>
            </Card.Body>
  
          </Card>
          )
      }
      else{
        return (
          <Card bg = "light" style={{ width: '28rem'}}>
            <Navbar bg="dark" variant="dark">
              <Container>
                <Nav bg="dark" className="justify-content-end" activeKey="/">
                  <Nav.Item>
                    <NavLink to={'/user/'+props.post.author}>
                      <ReactRoundedImage image={props.post.avatar} roundedColor="#ffffff" imageWidth="35" imageHeight="35" roundedSize="2" borderRadius="20" />
                    </NavLink>
                  </Nav.Item>
                  <Nav.Item> <NavLink to={'/user/'+props.post.author} className='headers'><strong>{props.post.author}</strong></NavLink> </Nav.Item>
                </Nav>
              </Container>
            </Navbar>
  
            <Card.Img variant="bottom" rounded="true" src={props.post.image} />
            <Card.Subtitle className="text-muted" style={{paddingTop: "1rem"}}> {props.post.date} </Card.Subtitle>
  
            <Card.Body>
              <Card.Text>{props.post.description} </Card.Text>
              <ButtonGroup aria-label="like,comment,message">
                  <Button variant="light">Like</Button>
                  <Button variant="light">Comment</Button>
              </ButtonGroup>
            </Card.Body>
          </Card>
          )
      }
    }

}



export default function FeedPage() {
  // get followings of the current user
    const stateCurrentUser = useSelector(selectCurrentUser);
    let followingsUsernames = stateCurrentUser.followings;
    const [feedPostOfUser, setFeedPostOfUser] = useState([]);
    const [popularPostsNearby, setPopularPostsNearby] = useState([]);

    useEffect(() => {
      async function fetchData() {
        const userObjects = await getUsersAsList(followingsUsernames);
        
        //remove user who has no post
        const userObjectsFiltered = userObjects.filter((object)=>{
            return object.posts.length !==0;
          });

        // map the object to it's latest post
        const followingsLatestPost = await Promise.all(userObjectsFiltered.map(async (o) => {
          //return the largest(most recent) ID
          let postObject = await getLastestPostOfAUser(o.username);
          //adding one more avatar field for this post
          postObject.avatar = o.avatar;
          return postObject;
        }))
        setFeedPostOfUser(followingsLatestPost);

        // fetch popular post
        // get the popular post 
        const popularPosts = await Promise.all([1,2,3].map(async (id) => {
          const post = await getPost(id);
          return post;
        }));
        setPopularPostsNearby(popularPosts);
      }
      try{
        fetchData();
      }
      catch(err){
          console.error(err);
      }
      
    }, []); //adding empty dependency making sure useEffect only run once after each render
  


  const cards = ((feedPostOfUser.length===0) ? popularPostsNearby:feedPostOfUser).map((p) => (
    <CardCustomed post={p}/>
  ))

  return (
    <div className='background'>
    
    <Stack gap={2} className="col-md-6 mx-auto">
        {cards}
    </Stack>
    
    </div>
  );
}