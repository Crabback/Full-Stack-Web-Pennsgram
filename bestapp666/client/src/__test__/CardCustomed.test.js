import '@testing-library/jest-dom/extend-expect';

import React from "react";
import { getByDisplayValue, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import userEvent from "@testing-library/user-event";
import { BrowserRouter } from "react-router-dom";
import FeedPage from "../pages/FeedPage/FeedPage";
import { Provider } from 'react-redux';
import { configureStore } from "@reduxjs/toolkit";
import store from '../Store/store';
import renderer from 'react-test-renderer';
import { updateCurrentUser, selectCurrentUser} from '../pages/UserPage/currentUserSlice';
import { getUser } from "../api/mock_api";
import { CardCustomed } from '../components/CardCustomed';

const store1 = store;

test('render feed page as logged in user list ', async () => {
    const currentUser = {username: 'obama', password: 'obama123'};
    const userRoster = await getUser(currentUser.username);
    let user;
    userRoster.forEach(element => {
      user = element;
    });
    store1.dispatch(updateCurrentUser(user));
    const component = renderer.create(
        <Provider store={store1}>
          <BrowserRouter>
            <CardCustomed />
          </BrowserRouter>
        </Provider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();

});
