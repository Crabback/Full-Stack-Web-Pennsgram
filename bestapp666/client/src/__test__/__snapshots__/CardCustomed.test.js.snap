// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render feed page as logged in user list  1`] = `null`;
